File: app/src/main/java/dev/ragnarok/fenrir/upload/IUploadManager.java
package dev.ragnarok.fenrir.upload;

import androidx.annotation.NonNull;

import java.util.List;

import dev.ragnarok.fenrir.util.Optional;
import dev.ragnarok.fenrir.util.Pair;
import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.core.Single;

public interface IUploadManager {
    Single<List<Upload>> get(int accountId, @NonNull UploadDestination destination);

    void enqueue(@NonNull List<UploadIntent> intents);

    void cancel(int id);

    void cancelAll(int accountId, @NonNull UploadDestination destination);

    Optional<Upload> getCurrent();

    Flowable<int[]> observeDeleting(boolean includeCompleted);

    Flowable<List<Upload>> observeAdding();

    Flowable<Upload> obseveStatus();

    Flowable<Pair<Upload, UploadResult<?>>> observeResults();

    Flowable<List<IProgressUpdate>> observeProgress();

    interface IProgressUpdate {
        int getId();

        int getProgress();
    }
}

File: app_filegallery/src/main/kotlin/dev/ragnarok/filegallery/upload/IUploadManager.kt
package dev.ragnarok.filegallery.upload

import dev.ragnarok.filegallery.util.Optional
import dev.ragnarok.filegallery.util.Pair
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.SharedFlow

interface IUploadManager {
    operator fun get(destination: UploadDestination): Flow<List<Upload>>
    fun enqueue(intents: List<UploadIntent>)
    fun cancel(id: Int)
    fun retry(id: Int)
    fun retryAll()
    fun cancelAll(destination: UploadDestination)
    fun getCurrent(): Optional<Upload>
    fun observeDeleting(includeCompleted: Boolean): Flow<IntArray>
    fun observeAdding(): SharedFlow<List<Upload>>
    fun observeStatus(): SharedFlow<Upload>
    fun observeResults(): SharedFlow<Pair<Upload, UploadResult<*>>>
    fun observeProgress(): Flow<IProgressUpdate?>
    interface IProgressUpdate {
        val id: Int
        val progress: Int
    }
}